src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/WebSocketStream.cs

```diff
@@ -0,0 +1,196 @@
+
+namespace System.Net.WebSockets
+{
+    /// <summary>Provides a <see cref="Stream"/> that delegates to a wrapped <see cref="WebSocket"/>.</summary>
+    public sealed class WebSocketStream : Stream
+    {
+        /// <summary>Whether disposing this instance should dispose <see cref="WebSocket"/>.</summary>
+        private readonly bool _ownsWebSocket;
+        /// <summary>Whether the instance has been disposed.</summary>
+        private bool _disposed;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="WebSocketStream"/> class using a specified <see cref="WebSocket"/> instance.
+        /// </summary>
+        /// <param name="webSocket">The <see cref="WebSocket"/> wrapped by this instance.</param>
+        /// <param name="ownsWebSocket">
+        /// <see langword="true"/> to indicate that the <see cref="WebSocketStream"/> takes ownership of the <see cref="WebSocket"/>,
+        /// such that disposing of this <see cref="Stream"/> will dispose of the <see cref="WebSocket"/>; otherwise, <see langword="false"/>.
+        /// When <see langword="true"/>, disposing of this instance doesn't initiate a close handshake; it merely delegates to
+        /// <see cref="WebSocket.Dispose"/>.
+        /// </param>
+        public WebSocketStream(WebSocket webSocket, bool ownsWebSocket = false)
+        {
+            ArgumentNullException.ThrowIfNull(webSocket);
+
+            WebSocket = webSocket;
+            _ownsWebSocket = ownsWebSocket;
+        }
+
+        /// <inheritdoc />
+        protected override void Dispose(bool disposing)
+        {
+            if (_disposed)
+            {
+                return;
+            }
+            _disposed = true;
+
+            if (disposing && _ownsWebSocket)
+            {
+                if (WebSocket.State is WebSocketState.Open)
+                {
+                    // There's no synchronous close, so we're forced to do sync-over-async.
+                    WebSocket.CloseOutputAsync(WebSocketCloseStatus.NormalClosure, null, CancellationToken.None).GetAwaiter().GetResult();
+                }
+
+                WebSocket.Dispose();
+            }
+
+            base.Dispose(disposing);
+        }
+
+        /// <inheritdoc />
+        public override ValueTask DisposeAsync()
+        {
+            if (_disposed)
+            {
+                return default;
+            }
+            _disposed = true;
+
+            if (_ownsWebSocket && WebSocket.State is WebSocketState.Open)
+            {
+                return CloseAndDisposeAsync();
+            }
+            else
+            {
+                return base.DisposeAsync();
+            }
+
+            async ValueTask CloseAndDisposeAsync()
+            {
+                if (WebSocket.State is WebSocketState.Open)
+                {
+                    await WebSocket.CloseOutputAsync(WebSocketCloseStatus.NormalClosure, null, CancellationToken.None).ConfigureAwait(false);
```

Comment:
Is it reasonable to use `CancellationToken.None` here? Since we're disposing the websocket if the remote party fails to respond to the Close request and we don't have a keep alive set, we will basically hang here.

Would it not be better if we have some timeout for the close to complete?

Reply:
Yes. The close timeout is one of the open questions at the moment. The full list is in https://github.com/dotnet/runtime/issues/111217#issuecomment-2813786887 -- we'd appreciate your thoughts on it as well 🙏

Reply:
From an API standpoint I would prefer if the close timeout can be specified in `WebSocketCreationOptions` much like the  KeepAliveTimeout. Doing so eliminates the need for the consumers of the `WebSocket` to know about it. I would even go further than this, making the `CloseTimeout` to have a meaningful value by default.